### PR TITLE
Support for flash attention

### PR DESCRIPTION
**Pull request by rebel-jindol21**
### 🚀 Pull Request Summary

Flash attention custom kernel is supported in vLLM.
Currently, this PR works along with triton(main branch) and rebel_compiler(vllm_moe_rsd_triton branch).

> *What does this PR do? What feature, fix, or improvement does it bring?*
Flash attention 
---

### 📌 Related Issues / Tickets

* Fixes #
* Related to #

---

### ✅ Changes

* [x] Feature
* [ ] Bug Fix
* [ ] Documentation
* [ ] Refactor
* [ ] Optimization
* [ ] Other (please describe):

---

### 📸 Screenshots / Logs (if applicable)

---

### 🧪 How to Test

1. Run `RBLN_KERNEL_MODE="triton" COMPILE_MODEL=1 python3 examples/experimental/qwen3_flash_test.py`
2. Verify output: 
  * Generated text: 'The companies will now work together to evaluate the performance of Rebellions’ ATOM-based NPU servers in SK Telecom’s NPU farm. This evaluation will help validate the performance and reliability of the technology in real-world, high-load environments. The results of this evaluation will be shared with the broader AI community, contributing to the development of more robust and scalable AI infrastructure.\n\nThe collaboration is expected to accelerate the development of next-generation AI infrastructure, enabling more efficient and sustainable AI systems. It also highlights the growing importance of AI infrastructure in the global tech landscape, as companies seek to build scalable, reliable, and energy-efficient AI solutions.\n\nRebellions, SK Telecom, and DOCOMO Innovations are now working together to advance the next generation of AI infrastructure. This partnership is a significant step toward the future of AI, and the companies are committed to driving innovation and growth in the AI hardware ecosystem.\n\n### \n\n**About Rebellions:**\nRebellions is a leading AI infrastructure company that provides high-performance AI chips and software solutions. The company is dedicated to advancing the global AI infrastructure market through innovation and collaboration.\n\n**About SK Telecom:**\nSK Telecom is a leading global telecommunications company that provides a wide range of services, including AI-optimized cloud services. The company'
  * Generated text: 'The companies will also explore opportunities in the AI data center market, including the development of AI infrastructure solutions, AI data center management solutions, and AI software stack delivery.\n\nThe collaboration is expected to drive innovation in AI infrastructure, reduce operational costs, and enhance the overall efficiency of AI data centers. This partnership is a significant step forward in the development of the global AI data center ecosystem, and the companies are committed to driving the future of AI infrastructure through continued collaboration and innovation.\n\nThe companies will also explore opportunities in the AI data center market, including the development of AI infrastructure solutions, AI data center management solutions, and AI software stack delivery.\n\nThe collaboration is expected to drive innovation in AI infrastructure, reduce operational costs, and enhance the overall efficiency of AI data center operations. This partnership is a significant step forward in the development of the global AI data center ecosystem, and the companies are committed to driving the future of AI infrastructure through continued collaboration and innovation.\n\nThe companies will also explore opportunities in the AI data center market, including the development of AI infrastructure solutions, AI data center management solutions, and AI software stack delivery.\n\nThe collaboration is expected to drive innovation in AI infrastructure, reduce operational costs, and enhance the overall efficiency of AI data center operations. This partnership is a significant step forward in'
4. Edge case tested: `...`

---

### 📋 Checklist

* [ ] I’ve added unit tests or updated existing ones.
* [ ] I’ve updated relevant documentation.
* [ ] I’ve checked compatibility (e.g., hardware/backend/platform).

---

### 💬 Notes

---

### CODE REVIEW

File: vllm_rbln/worker/worker.py [hunk 1]
```diff
@@ -289,9 +290,16 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
         # sequences that can be processed in a single batch. This is equivalent
         # to schedule without PagedAttention.
 
-        # NOTE(jiwoo.park):
-        # We reserve one shared cache block for padded batch sequences.
-        num_gpu_blocks = self.scheduler_config.max_num_seqs + 1
+        # self.model_config or self.scheduler_config or self.vllm_config
+        block_size = self.model_runner.block_size
+        max_num_seqs = self.model_runner.scheduler_config.max_num_seqs
+        max_model_len = self.model_runner.scheduler_config.max_model_len
```
Comment (rebel-jiwoopark) [inline vllm_rbln/worker/worker.py:296]: Would it be possible to use the Worker's `scheduler_config` and `cache_config`?

**Comment by rebel-eunji** [comment]
Sorry, I tried running this code and got an error in rebel-compiler (on the `vllm_moe_rsd_triton` branch).
Does the compiler require PyTorch?
```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
C10_LIB
    linked by target "rbln_runtime_objs" in directory /home/eunji.lee/plugin/rebel_compiler
TORCH_CPU_LIB
    linked by target "rbln_runtime_objs" in directory /home/eunji.lee/plugin/rebel_compiler

-- Generating done
```

File: vllm_rbln/worker/worker.py [hunk 2]
```diff
@@ -289,9 +290,16 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
...
+
+        # no of blocks is determined by user-defined value or model spec
+        # -1 : this means that we need a last block as dummy block.
+        num_gpu_blocks = math.ceil(max_model_len / block_size) * max_num_seqs
```
Comment (rebel-jonghewk) [inline vllm_rbln/worker/worker.py:300]: This assumption may not always hold true, as the actual available DRAM could be less than what is implied by `num_gpu_blocks`. While vLLM estimates `num_gpu_blocks` using a dummy run, in our case, we should follow the estimation approach used in Optimum-RBLN instead [[reference](https://github.com/rebellions-sw/optimum-rbln/blob/main/src/optimum/rbln/transformers/models/decoderonly/modeling_decoderonly.py#L770)].

Comment (rebel-jindol21) [reply vllm_rbln/worker/worker.py:300]: Thank you for comments. I will look into it and put it int vLLM.

**Comment by rebel-jindol21** [comment]
More tests are required. This will be closed and a new patch will be soon.

File: vllm_rbln/worker/worker.py (hunk 1 above)
Comment (rebel-jiwoopark) [reply vllm_rbln/worker/worker.py]: Thank you. I checked the changes.

File: vllm_rbln/worker/worker.py [hunk 3]
```diff
@@ -289,9 +290,25 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
...
+            nbits_per_param=16 if not self.model_config.quantization else 4,
+            n_model_params=sum(p.numel()
+                               for p in self.model_runner.model.parameters()),
+            # 1 : prefill
+            num_runtimes=1 + self.scheduler_config.max_num_seqs)
+        num_gpu_blocks = (max_num_blocks - 1)
+
+        if npu_num_blocks := os.environ.get("VLLM_RBLN_NPU_NUM_BLOCKS"):
+            num_gpu_blocks = int(npu_num_blocks) - 1
+        num_cpu_blocks = 0
```
[the remark this answers is not in the document]
Comment (rebel-jindol21) [reply vllm_rbln/worker/worker.py:311]: Resolved. Thanks.

[the remark this answers is not in the document]
Comment (rebel-jiwoopark) [reply vllm_rbln/worker/worker.py]: Confirmed, thanks!

File: examples/experimental/qwen3_flash_test.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    model=qwen3_4_model_id,
+    max_model_len=16 * 128,
+    #max_model_len=2048,
+    block_size=128,
+    #block_size=1024,
+    enable_chunked_prefill=True,
+    max_num_batched_tokens=128,
+    max_num_seqs=5,
+    #tensor_parallel_size=4,
+    #enable_expert_parallel=True,
```
Comment (rebel-jiwoopark) [inline examples/experimental/qwen3_flash_test.py:80]: It's not a major concern, but you may want to consider tidying up any unused models and options.

Comment (rebel-jindol21) [reply examples/experimental/qwen3_flash_test.py:80]: They were left for the reference. I will erase unused arguments.

File: vllm_rbln/worker/model_runner.py [hunk 5]
```diff
@@ -296,10 +309,18 @@ def _prepare_decode(
...
 
         # batch padding
-        batch_padding_szie = self.max_num_seqs - len(data.input_tokens)
-        data.input_tokens.extend([[0]] * batch_padding_szie)
-        data.input_positions.extend([[0]] * batch_padding_szie)
-        arr_input_block_ids.extend([self.max_num_seqs] * batch_padding_szie)
+        batch_padding_size = self.max_num_seqs - len(data.input_tokens)
+        data.input_tokens.extend([[0]] * batch_padding_size)
+        data.input_positions.extend([[0]] * batch_padding_size)
+        list_input_block_ids.extend([[0]] * batch_padding_size)
```
[the remark this answers is not in the document]
Comment (rebel-jindol21) [reply vllm_rbln/worker/model_runner.py]: To make it simple, it would be good that all block ids become "dummy" id.

File: vllm_rbln/worker/worker.py (hunk 2 above)
Comment (rebel-jiwoopark) [reply vllm_rbln/worker/worker.py]: @rebel-jonghewk PTAL

File: vllm_rbln/worker/worker.py [hunk 6]
```diff
@@ -287,9 +288,24 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
...
-        # NOTE(jiwoo.park):
-        # We reserve one shared cache block for padded batch sequences.
-        num_gpu_blocks = self.scheduler_config.max_num_seqs + 1
+        block_size = self.cache_config.block_size
+
+        # This function comes from optimum-rbln.
+        # We must keep it updated as optimum is upgraded.
+        max_num_blocks = get_maximum_num_blocks(
+            config=self.model_config,
+            tensor_parallel_size=self.parallel_config.tensor_parallel_size,
```
Comment (rebel-jonghewk) [inline vllm_rbln/worker/worker.py:297]: Just to confirm — is `parallel_config.tensor_parallel_size` equivalent to the tensor_parallel_size used in RSD?
